lib/kube/proxy/pod_filters.go

```diff
@@ -0,0 +1,306 @@
+		filtered = obj
+	case *corev1.Pod:
+		// filterPod filters a single corev1.Pod and returns an error if access to
+		// pod was denied.
+		if err := filterPod(o, d.allowedResources, d.deniedResources); err != nil {
+			if !trace.IsAccessDenied(err) {
+				d.log.WithError(err).Warnf("Unable to compile role kubernetes_resources.")
+			}
+			return trace.Wrap(err)
+		}
+	case *corev1.PodList:
+		filtered = filterCoreV1PodList(o, d.allowedResources, d.deniedResources, d.log)
+	case *metav1.Table:
+		filtered, err = d.filterMetav1Table(o, d.allowedResources, d.deniedResources)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	default:
+		return trace.BadParameter("unexpected type received; got %T", obj)
+	}
+	// encode the filterer response back to the user.
+	return d.encode(filtered, output)
+}
+
+// FilterObj receives a runtime.Object type and filters the resources on it
+// based on allowed and denied rules resources.
+// After filtering them, the obj is manipulated to hold the filtered information.
+func (d *podFilterer) FilterObj(obj runtime.Object) (bool, error) {
+	switch o := obj.(type) {
+	case *corev1.Pod:
+		err := filterPod(o, d.allowedResources, d.deniedResources)
+		if err != nil && !trace.IsAccessDenied(err) {
+			d.log.WithError(err).Warnf("Unable to compile role kubernetes_resources.")
+		}
+		// if err is not nil we should not include it.
+		return err == nil, nil
+	case *corev1.PodList:
+		_ = filterCoreV1PodList(o, d.allowedResources, d.deniedResources, d.log)
+		return len(o.Items) > 0, nil
+	case *metav1.Table:
+		_, err := d.filterMetav1Table(o, d.allowedResources, d.deniedResources)
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+		return len(o.Rows) > 0, nil
+	default:
+		return false, trace.BadParameter("unexpected type received; got %T", obj)
+	}
+}
+
+// decode decodes the buffer into the appropriate type if the responseCode is valid.
+// If not valid, returns the buffer unchanged.
+func (d *podFilterer) decode(buffer []byte) (runtime.Object, []byte, error) {
+	switch {
+	case d.responseCode == http.StatusSwitchingProtocols:
+		// no-op, we've been upgraded
+		return nil, buffer, nil
+	case d.responseCode < http.StatusOK /* 200 */ || d.responseCode > http.StatusPartialContent /* 206 */ :
+		// calculate an unstructured error from the response which the Result object may use if the caller
+		// did not return a structured error.
+		// Logic from: https://github.com/kubernetes/client-go/blob/58ff029093df37cad9fa28778a37f11fa495d9cf/rest/request.go#L1040
+		return nil, buffer, nil
+	default:
+		out, err := decodeAndSetGVK(d.decoder, buffer)
+		return out, nil, trace.Wrap(err)
+	}
+}
+
+// decodePartialObjectMetadata decodes the metav1.PartialObjectMetadata present
+// in the metav1.TableRow entry. This information comes from server side and
+// includes the resource name and namespace as a structured object.
+func (d *podFilterer) decodePartialObjectMetadata(row *metav1.TableRow) (err error) {
```

Comment:
nit: uninteresting named return

```suggestion
func (d *podFilterer) decodePartialObjectMetadata(row *metav1.TableRow) error {
```